[feat] id, password를 DB에서 확인한 후 토큰 반환하는 기능 구현

j03y14 opened the pull request:
# id, password를 DB에서 확인한 후 토큰 반환하는 기능 구현
## 관련 이슈 및 위키
**이슈**: #9 #10 

## 내용
### passport stretegy 구현
> 5db3caf

각 passport strategy에는 config 객체와 verify 함수가 들어갑니다.
```
- config: stretegy의 옵션을 설정
- verify: 해당하는 유저가 있는지 확인
```
verify 함수에서 social 값은 자동으로 넘어오지 않기 때문에 passportVerify 함수를 클로저로 만들고 social 값을 추가해서 각 소셜 로그인에 맞는 verify 함수를 추가하도록 했습니다.
### select user query 추가
> 631c27a

username과 social 값을 가지고 있는 유저가 있는지를 확인합니다.
### authModel 중 select 구현
> 5c95d82

### authController 중 login 구현
> 30b29d6

`POST /api/auth/:social` api 요청이 들어오면  social 값에 따라서 다른 passport 전략을 사용해서 인증을 진행합니다.

### auth router 중 login 구현
> 068be11

`POST /api/auth/:social` api 요청을 받습니다.

## Why?
### authModel 
> 5c95d82

```javascript
const [res] = await connection.query(sql.selectUser, [username, social]);
return res[0];
```
- query에서 반환되는 배열의 첫번째 요소만 실질적인 데이터가 들어가 있는 반환값이라서 `[res]`로 받았습니다.
- res는 객체 하나만 담긴 배열이라서 `res[0]`으로 return 했습니다.

j03y14 committed:
[feat] passport 추가 및 에러 메세지 출력 미들웨어 추가

passport 설정을 하는 passportConfig()를 추가하고 어느 경로에도 포함이 되지 않는 요청이 들어왔을 때 에러를 출력하는 부분을 추가했습니다.

j03y14 committed:
[feat] passport github, apple 소셜 로그인에 대항하는 유저가 있는지 확인하는 기능 추가

각 passport strategy에는 config 객체와 verify 함수가 들어갑니다.

config에서 stretegy의 옵션을 설정하고 verify에서 해당하는 유저가 있는지 확인합니다.

verify 함수에서 social 값은 자동으로 넘어오지 않기 때문에 passportVerify 함수를 클로저로 만들고 social 값을 추가해서 각 소셜 로그인에 맞는 verify 함수를 추가하도록 했습니다.

j03y14 committed:
[feat] username과 social로 select를 하는 query 추가

username과 social 값이 모두 맞는 유저가 있는지 확인을 하는 과정이 로그인에서 필요하기 때문에 query를 추가했습니다.

j03y14 committed:
[feat] authModel 중 select 구현

username과 social을 가지고 DB에 유저가 있는지 확인을 합니다.
query 결과 배열의 첫번째 항만 select의 result에 해당하므로
const [res]
로 비구조화 해서 필요한 값만 받았습니다.
그 이후에 select의 결과는 배열이므로 res[0]만 return 했습니다.

j03y14 committed:
[chore] http-error, cookie-parser, mysql2, passport, passport-jwt, passport-local 패키지 추가

j03y14 committed:
[feat] authController의 login 메소드 구현

social에 해당하는 값을 param으로 받아서 `${social}Local`에 해당하는 strategy로 인증을 진행하고 인증이 되면 토큰을 생성해서 응답을 보냅니다.

j03y14 committed:
[feat] auth의 로그인 api 생성

POST /api/auth/[social]에 해당하는 api를 만들었습니다. social을 가지고 어떤 stretegy로 인증을 하는지가 결정이 됩니다.

safience-mia commented on web-BE/app.js:
```diff
@@ -1,15 +1,37 @@
+
+app.use((err, req, res, next) => {
+  res.locals.message = err.message;
+  res.locals.error = req.app.get('env') === 'development' ? err : {};
```
제가 따로 학습을 해야겠지만 
`http-errors`를 사용해서 에러 관리하는 방법론 설명을 한번 해주시면 좋겠어요!
아주 후루룩 훑은 바로는 `createError(404)`로 에러 객체가 생성된다고 하는데 
지금 코드로는 확실히 이해가 안가요 🥺 
이 방법 괜찮다면 팀 수준으로 채택에서 사용하면 좋겠네요!

j03y14 replied:
저도 막 알고 한건 아니라서 잘 모르겠네요ㅠㅠ
https://libsora.so/web/express-error-and-http-status-code/
이런 자료를 찾긴 했는데 제가 좀 더 알아보고 말씀드릴게요

SeunghyoKu replied:
수원님, 재윤님 말씀 듣고 찾아봤는데 오류가 발생하면 예외 처리를 하는 미들웨어를 이용하는 방식이군요 👀 신기합니다.
https://www.npmjs.com/package/http-errors

npm에서 약간 찾아봤는데 `http-errors`를 이용하면 굉장히 간단하게 에러를 만들 수 있나봐요.

safience-mia commented on web-BE/config/passport_config.js:
```diff
@@ -0,0 +1,48 @@
+
+const jwtConfig = {
+  jwtFromRequest: ExtractJWT.fromAuthHeaderAsBearerToken(),
+  secretOrKey: process.env.JWT_SECRET,
```
JWT_SECRET 값 공유해주세용!

j03y14 replied:
여기 남기면 안 될 것 같아서 월요일에 줌으로 말씀드릴게요😀

safience-mia commented on web-BE/config/query.js:
```diff
@@ -0,0 +1,8 @@
+  selectLabel: 'SELECT * FROM Label',
+  updateLabel: 'UPDATE Label SET name=?, description=?, color=?, WHERE label_id=?;',
+  deleteLabel: 'DELETE FROM Label WHERE label_id=?;',
+  selectUser: 'SELECT user_id, username, password, social FROM User WHERE username=? and social=?',
```
(코드제안하는 기능이 있던데 어떻게 하지..? 그러면 바로 고쳐서 commit/push 할 수 있더라구요)

아주 사소하지만 구분 주석 있으면 좋겠어융 🌝
```javascript
 // user
selectUser: 'SELECT user_id, username, password, social FROM User WHERE username=? and social=?',
```

j03y14 replied:
넵 주석은 추가하는게 좋을 것 같네요!

SeunghyoKu replied:
이 부분이 제 코드랑 충돌할 것 같군요..! 머지 충돌을 해결해본 적이 없어서 내일 함께 해결해보아야할 것 같아요.  😅

safience-mia commented on web-BE/config/query.js:
```diff
@@ -0,0 +1,8 @@
+  selectLabel: 'SELECT * FROM Label',
+  updateLabel: 'UPDATE Label SET name=?, description=?, color=?, WHERE label_id=?;',
+  deleteLabel: 'DELETE FROM Label WHERE label_id=?;',
+  selectUser: 'SELECT user_id, username, password, social FROM User WHERE username=? and social=?',
```
- 유저를 select 할 때 password 가 필요할까요? 로그인은 SNS로그인하고, 우리한테 들어오는 유저인포는 패스워드를 포함하지 않을테니까요! 우리 서버에서 회원가입 처리를 할 때에도, 패스워드는 필요하지 않을 것 같아요 🧐
- SNS 로그인 -> DB에 없는 유저인 경우 : 회원가입 
  -> 가입 절차 시 : 비밀번호 등록 없이, SNS OAuth로부터 제공받은 유저인포 확인, 정보 제공 동의 정도로 진행 되지 싶어요 
  아래는 참고용으로 테스트 해본 SNS register 절차입니다! 
  혼동하기 쉽고 중요한 부분이니 프로세스 로직을 웹/iOS 회의 통해 문서화하고 가도 좋을 것 같아요
  <img src='https://user-images.githubusercontent.com/45927473/97693972-49a03080-1ae5-11eb-86ba-050eed4eac7a.gif' width=450px >

- 만약 패스워드가 필요없는 게 맞다면, user 테이블의 password 칼럼은 삭제해야겠네요 😦

j03y14 replied:
password를 가져올 필요는 없겠네요 정말!

j03y14 replied:
이게 생각해보니까 약간 이상한 부분도 있는데 월요일에 줌으로 말씀드릴게요

safience-mia commented on web-BE/models/auth.js:
```diff
@@ -0,0 +1,16 @@
+  select: async (username, social) => {
+    try {
+      const [res] = await connection.query(sql.selectUser, [username, social]);
+      return res[0];
```
- 커밋메세지 `query 결과 배열의 첫번째 항만 select의 result에 해당하므로` 에서 `sql.selectUser` 의 결과로 단일 Array말고 더 오는 경우가 있나요? 
- 좀 더 직관적으로 질문 드리자면, `const [res, tmp]` 로 정의 했을 때 `tmp` 에 할당되는 무언가가 있나요? 
- 질문 드린 이유는 제가 코드 작성을 잘못 했을 수 도 있기 때문인데요,,, 🙄
  - 제 커밋 5951f68 에서 보시면 아래와 같이 작성 해놨거든요, 문제가 있을까요? 
  - ```javascript 
      const res = await connection.query(sql.selectLabel, labelid);
      return res[0]; 
    ```

j03y14 replied:
https://www.npmjs.com/package/mysql2
```javascript
async function main() {
  // get the client
  const mysql = require('mysql2/promise');
  // create the connection
  const connection = await mysql.createConnection({host:'localhost', user: 'root', database: 'test'});
  // query database
  const [rows, fields] = await connection.execute('SELECT * FROM `table` WHERE `name` = ? AND `age` > ?', ['Morty', 14]);
}
```
위 링크에 있는 예제인데 rows랑 fields가 나온다고 해요. fields를 어디에 쓰는건지는 잘 모르겠어요😥
실제로 log를 출력해봤는데
```javascript
[
  [
    TextRow {
      user_id: 2,
      username: 'appletest',
      password: 'test',
      social: 'apple'
    }
  ],
  [
    ColumnDefinition {
      _buf: <Buffer 01 00 00 01 04 38 00 00 02 03 64 65 66 0c 49 73 73 75 65 54 72 61 63 6b 65 72 04 55 73 65 72 04 55 73 65 72 07 75 73 65 72 5f 69 64 07 75 73 65 72 5f ... 242 more bytes>,
      _clientEncoding: 'utf8',
      _catalogLength: 3,
      _catalogStart: 10,
      _schemaLength: 12,
      _schemaStart: 14,
      _tableLength: 4,
      _tableStart: 27,
      _orgTableLength: 4,
      _orgTableStart: 32,
      _orgNameLength: 7,
      _orgNameStart: 45,
      characterSet: 63,
      encoding: 'binary',
      name: 'user_id',
      columnLength: 11,
      columnType: 3,
      flags: 16899,
      decimals: 0
    },
    ColumnDefinition {
      _buf: <Buffer 01 00 00 01 04 38 00 00 02 03 64 65 66 0c 49 73 73 75 65 54 72 61 63 6b 65 72 04 55 73 65 72 04 55 73 65 72 07 75 73 65 72 5f 69 64 07 75 73 65 72 5f ... 242 more bytes>,
      _clientEncoding: 'utf8',
      _catalogLength: 3,
      _catalogStart: 70,
      _schemaLength: 12,
      _schemaStart: 74,
      _tableLength: 4,
      _tableStart: 87,
      _orgTableLength: 4,
      _orgTableStart: 92,
      _orgNameLength: 8,
      _orgNameStart: 106,
      characterSet: 33,
      encoding: 'cesu8',
      name: 'username',
      columnLength: 135,
      columnType: 253,
      flags: 4097,
      decimals: 0
    },
    ColumnDefinition {
      _buf: <Buffer 01 00 00 01 04 38 00 00 02 03 64 65 66 0c 49 73 73 75 65 54 72 61 63 6b 65 72 04 55 73 65 72 04 55 73 65 72 07 75 73 65 72 5f 69 64 07 75 73 65 72 5f ... 242 more bytes>,
      _clientEncoding: 'utf8',
      _catalogLength: 3,
      _catalogStart: 132,
      _schemaLength: 12,
      _schemaStart: 136,
      _tableLength: 4,
      _tableStart: 149,
      _orgTableLength: 4,
      _orgTableStart: 154,
      _orgNameLength: 8,
      _orgNameStart: 168,
      characterSet: 33,
      encoding: 'cesu8',
      name: 'password',
      columnLength: 135,
      columnType: 253,
      flags: 4097,
      decimals: 0
    },
    ColumnDefinition {
      _buf: <Buffer 01 00 00 01 04 38 00 00 02 03 64 65 66 0c 49 73 73 75 65 54 72 61 63 6b 65 72 04 55 73 65 72 04 55 73 65 72 07 75 73 65 72 5f 69 64 07 75 73 65 72 5f ... 242 more bytes>,
      _clientEncoding: 'utf8',
      _catalogLength: 3,
      _catalogStart: 194,
      _schemaLength: 12,
      _schemaStart: 198,
      _tableLength: 4,
      _tableStart: 211,
      _orgTableLength: 4,
      _orgTableStart: 216,
      _orgNameLength: 6,
      _orgNameStart: 228,
      characterSet: 33,
      encoding: 'cesu8',
      name: 'social',
      columnLength: 135,
      columnType: 253,
      flags: 0,
      decimals: 0
    }
  ]
]
```
이렇게 나오더라고요

safience-mia commented on web-BE/config/passport_config.js:
```diff
@@ -0,0 +1,48 @@
+  if (!user) {
+    return done(null, false);
+  }
+  return done(null, user);
```
`done()` 설명 해 주실 재윤님 구함 😭 
재윤님 PR에서 물음표 살인마가 된 느낌이네요 ㅠㅠ 
어떤 역할인지 설명 부탁드립니다 ㅠㅠ

j03y14 replied:
passport에서 **인증 결과를 정해주는 함수**에요. 
passport 내부에 구현되어 있는 거라고 보시면 될 것 같아요.
```javascript
// done([서버내부에러], [db조회 성공시 반환되는 데이터값], [에러메세지])
```
여기에 인자가 3개가 들어갈 수 있는데,

첫번째는 `서버의 에러`가 생겼을 때 에러를 넣어주는 인자이고,
두번째는 로그인시에 `user 객체`를 넣어주는 인자이고,
세번째는 `메세지 객체`를 넣어주면 된대요

그런데 만약에 DB에서 유저를 조회했을 때 없으면 서버에서 에러가 있는건 아니고 조회된 값이 없으니까 
```javascript
return done(null, false);
```
이렇게 넣어주고,
조회된 유저가 있으면 유저 정보를 넘겨주려고
```javascript
return done(null, user);
```
이렇게 했습니다.

SeunghyoKu replied:
저도 이 부분을 몰라서 찾아봤는데 (뭔가 위에도 똑같은 말을 남긴 거 같은데... 느낌 탓이 아닙니다.)
```text
If the credentials are valid, the verify callback invokes 'done' to supply Passport with the user that authenticated.
```
자격증명이 유효할 경우에 done에 다가 유효한 값을 passport에게 제공하는 방식인가봅니다. 재윤님의 친절한 설명에 감사드리는..22

safience-mia commented on web-BE/routes/api/auth/index.js:
```diff
@@ -0,0 +1,6 @@
+const router = require('express').Router();
+const authController = require('./controller');
+
+router.post('/:social', authController.login);
```
제 PR https://github.com/boostcamp-2020/IssueTracker-12/pull/51#issue-512046411 의 Why? 부분에서 한번 언급 한 부분이긴 한데, API 리퀘스트 방법을 한번 정리하고 가야 할 것 같습니다. 
 - 어떤 API에 어떤 리퀘스트 형식을 갖출지!

- 코드 컨벤션을 필요 이상으로 맞추려하지 않도록 의식하고 있습니다만 
  - 이부분은 개발자의 성향보다는 때에 따라 적절한 리퀘스트 방법이 있다고 생각해서 학습 겸 제안드려요..! 
  - (면접 질문거리로도 넘나 나올만하지 않나요..? 👀 이 상황엔 어떤 API 리퀘가 알맞다고 생각하나요? 왜그렇게 생각하죠? 등..)
- 혹시 과하다 생각드신다면 말씀해주세요!! 어차피 API 문서 작성할거고, 문서에 명시된대로 사용하면 그만이니까요 🌝

j03y14 replied:
크게 생각을 해보지 않고 어제 수원님 PR을 보고 그렇게 해야지 하고 있었는데 제가 의식을 못하고 코딩을 했네요😅

그래서 어떤게 더 좋은걸까, 기준이 뭘까 하고 찾아봤더니
각각 사용해야 하는 상황이 다음과 같다고 해요
- query string: 정렬하거나 필터해서 보여줘야 할 경우
- path variable: 어떤 자원(데이터)의 위치를 특정해서 보여줘야 할 경우

그래서 오히려 수원님께서 delete 하실 때 id는 query string으로 쓰고 제가 한 거는 query string이 맞을지도 모르겠네요

SeunghyoKu replied:
부가적으로 조금 더 찾아본 결과를 덧붙이자면,
- path parameter: 특정 리소스를 정의할 필요가 있을 때
- query parameter: 정렬 혹은 필터링이 필요할 때

그러니까 음.. delete할 때는 `path string`을 이용하고 이 social 구분에는 `query string`을 사용하는 것이 좋겠다는 생각이 들어요.
이 부분에 대해서도 다음에 이야기해보아요!

slogup-mia reviewed (approved):
- 쓰다보니 코멘트가 매우 많아졌습니다,, 수정 제안보다는 질문이 많습니다,, 부디 너른 마음으로 피드백 해주시길 바라요 
- @SeunghyoKu  승효님과 재윤님 이번 PR 겹치는게 꽤 있어보입니다,, 
  - 늦게 머지하시는 분에게 고난이도 컨플릭트 처리 실습 기회를 얻을 축복이 있나니.. 🙏🏻
  - 아니면 두분 가위바위보 하세요 ㅋㅋㅋㅋ

SeunghyoKu reviewed (approved):
리뷰가 죄다 공감하는 내용 밖에 없게 되었군요. 😭
네이밍과 코드 흐름도 명확해서 보기 좋았던 것 같아요.
모르는 내용이 많아서 수원님, 재윤님 말씀하셨던 부분 찾아보면서
나름 도움이 될 만한 부분을 덧붙여보았습니다. 월요일에 뵈어요.. 👍

j03y14 committed:
[fix] 인증을 해서 토큰을 만드는 작업 삭제

인증을 해서 토큰을 만드는 작업은 프론트엔드로 빼내는 것이 맞다고 판단해서 삭제했습니다.

j03y14 committed:
[fix] query 구분 주석을 추가

query를 구분할 수 있는 주석을 추가했습니다.

j03y14 committed:
[fix] login api를 user check api로 변경

실제로 로그인을 해서 토큰을 만드는 과정을 프론트엔드에서 하도록 하고 백엔드에서는 유저가 있는지 없는지 체크만 하도록 변경했습니다.
따라서 함수명도 login에서 checkUser로 변경했습니다.

j03y14 commented:
## 변경사항
### sql 주석 추가
> 55aa8fd

sql에 user와 관련된 query라는 것을 알려주기 위한 주석을 추가했습니다.

### 인증 후 토큰을 만드는 작업 삭제
> 16cf47d

인증 후 토큰을 만드는 작업은 프론트엔드에서 하도록 변경하고 백엔드에서는 토큰을 파싱해서 인증을 하는 과정과 username, social에 해당하는 유저가 있는지 확인만 하도록 변경했습니다. social 로그인을 할 때 비밀번호를 알아낼 수 없기 때문에 아무나 api 요청을 보내서 데이터를 가져갈 수 있는 위험이 있다고 생각했기 때문에 이렇게 변경했습니다.

j03y14 committed:
Merge branch 'web-BE' into web-BE-재윤-login
